File: uisdk/src/main/java/com/karhoo/uisdk/base/BasePresenter.java
package com.karhoo.uisdk.base;

import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;

public abstract class BasePresenter<V> {

    private WeakReference<V> viewReference;

    protected BasePresenter() {

    }

    @Nullable
    public V getView() {
        return viewReference == null ? null : viewReference.get();
    }

    public void attachView(V view) {
        viewReference = new WeakReference<>(view);
    }

    public void detachView() {
        if (viewReference != null) {
            viewReference.clear();
            viewReference = null;
        }
    }

    protected boolean viewIsAttached() {
        return viewReference != null && viewReference.get() != null;
    }

}

File: app/src/main/java/com/karhoo/karhootraveller/presentation/base/BasePresenter.kt
package com.karhoo.karhootraveller.presentation.base

import java.lang.ref.WeakReference

@Suppress("UnnecessaryAbstractClass")
abstract class BasePresenter<V> protected constructor() {

    private var viewReference: WeakReference<V>? = null

    val view: V?
        get() = if (viewReference == null) null else viewReference!!.get()

    fun attachView(view: V) {
        viewReference = WeakReference(view)
    }

    fun detachView() {
        if (viewReference != null) {
            viewReference!!.clear()
            viewReference = null
        }
    }

    protected fun viewIsAttached(): Boolean {
        return viewReference != null && viewReference!!.get() != null
    }

}
